### DISCUSSION TITLE
symbol

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#82] [프론트 기술 질문 공간]
자바스크립트의 자료형 중 symbol에 대해서 설명해주세요

**Comment by michoball** [comment]
symbol은 es6에서 새롭게 추가된 원시 타입입니다. 

symbol은 유일성이 보장됩니다. -> 같은 모습의 symbol이라도 항상 같지 않다.
 때문에 객체의 key값으로 사용될 수 있습니다.
 하지만 문자열과 다르게 symbol은 숨김 프로퍼티라 외부에서 객체에 접근할 수 없습니다.
 
symbol을 생성할 때는 매개변수를 넣을 수 있습니다. 하지만 이 매개변수는 해당 symbol에 대한 설명 정도의 역할을 할 뿐이라 굳이 설명이 필요없다면 매개변수에 문자열을 넣지 않아도 됩니다.

**Comment by WooGie911** [comment]
자바스크립트는 객체 프로퍼티 키로 오직 문자형과 심볼형만을 허용합니다. 
숫자형, 불린형 모두 불가능하고 오직 문자형과 심볼형만 가능합니다.

그 중 symbol은 유일한 식별자(unique identifier)를 만들고 싶을 때 사용합니다.
Symbol()을 사용하면 심볼값을 만들 수 있으며 심볼을 만들 때 심볼 이름이라 불리는 설명을 붙일 수도 있는데,
심볼에 붙이는 설명(심볼 이름)은 어떤 것에도 영향을 주지 않는 이름표 역할만을 합니다. 
또한 심볼은 유일성이 보장되는 자료형이기 때문에, 설명이 동일한 심볼을 여러 개 만들어도 각 심볼값은 다릅니다.
(심볼 이름이 같은 두 심볼을 동일 연산자(==)으로 비교해도 false가 반환됩니다.)

심볼을 이용하면 ‘숨김(hidden)’ 프로퍼티를 만들 수 있습니다. 
숨김 프로퍼티는 외부 코드에서 접근이 불가능하고 값도 덮어쓸 수 없는 프로퍼티입니다.
따라서 Object.keys(user)에서도 키가 심볼인 프로퍼티는 배제됩니다. 
또한 '심볼형 프로퍼티 숨기기(hiding symbolic property)'라 불리는 이런 원칙 덕분에 외부 스크립트나 라이브러리는 심볼형 키를 가진 프로퍼티에 접근하지 못합니다.

**Comment by javascriptbiggosoo** [comment]
Symbol은 원시형 데이터로, 유일무이한 식별자를 만드는 데 사용됩니다.

심볼의 주요 유스 케이스는 다음과 같습니다.

1. 객체의 ‘숨김’ 프로퍼티 – 외부 스크립트나 라이브러리에 ‘속한’ 객체에 새로운 프로퍼티를 추가해 주고 싶다면 심볼을 만들고, 이를 프로퍼티 키로 사용하면 됩니다. 키가 심볼인 경우엔 for..in의 대상이 되지 않아서 의도치 않게 프로퍼티가 수정되는 것을 예방할 수 있습니다. 외부 스크립트나 라이브러리는 심볼 정보를 갖고 있지 않아서 프로퍼티에 직접 접근하는 것도 불가능합니다. 심볼형 키를 사용하면 프로퍼티가 우연히라도 사용되거나 덮어씌워 지는 걸 예방할 수 있습니다.
이런 특징을 이용하면 원하는 것을 객체 안에 ‘은밀하게’ 숨길 수 있습니다. 외부 스크립트에선 우리가 숨긴 것을 절대 볼 수 없습니다.

2. 자바스크립트 내부에서 사용되는 시스템 심볼은 Symbol.*로 접근할 수 있습니다. 

사실 심볼을 완전히 숨길 방법은 없습니다. 내장 메서드 `Object.getOwnPropertySymbols(obj)`를 사용하면 모든 심볼을 볼 수 있고, 메서드 `Reflect.ownKeys(obj)`는 심볼형 키를 포함한 객체의 모든 키를 반환해줍니다. 그런데 대부분의 라이브러리, 내장 함수 등은 이런 메서드를 사용하지 않습니다.
